src/query/service/src/pipelines/executor/queries_pipeline_executor.rs

```diff
@@ -413,16 +182,30 @@ impl QueriesPipelineExecutor {
             }
 
             while !self.global_tasks_queue.is_finished() && context.has_task() {
-                if let Some((executed_pid, graph)) = context.execute_task(Some(self))? {
-                    // Not scheduled graph if pipeline is finished.
-                    if !self.global_tasks_queue.is_finished() {
-                        // We immediately schedule the processor again.
-                        let schedule_queue = graph.schedule_queue(executed_pid)?;
-                        schedule_queue.schedule_with_condition(
-                            &self.global_tasks_queue,
-                            &mut context,
-                            self,
-                        );
+                match context.execute_task(Some(self)) {
+                    Ok(Some((executed_pid, graph))) => {
+                        // Not scheduled graph if pipeline is finished.
+                        if !self.global_tasks_queue.is_finished() && !graph.is_should_finish() {
+                            // We immediately schedule the processor again.
+                            let schedule_queue = graph.clone().schedule_queue(executed_pid)?;
+                            schedule_queue.schedule_with_condition(
+                                &self.global_tasks_queue,
+                                &mut context,
+                                self,
+                            );
+                        }
+                        if graph.is_should_finish() {
+                            // TODO: temporary finish method, will remove after change executor to a global service
+                            self.finish(None);
+                        }
+                        if graph.is_all_nodes_finished() {
+                            graph.should_finish(Ok(()))?;
+                        }
+                    }
```

Comment:
I am not confident about this part. /cc @zhang2014 
I tried to separate `query's error` and `executor's error`. Do you have any comment?